core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java

```diff
@@ -0,0 +1,224 @@
+        this.scoreRingBuffer = scoreRingBuffer;
+        this.readIndex = readIndex;
+        this.writeIndex = writeIndex;
+    }
+
+    record RingBufferState(int readIndex, int writeIndex,
+            long[] nanoTimeRingBuffer, @Nullable Score<?>[] scoreRingBuffer) {
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof RingBufferState that)) {
+                return false;
+            }
+            return readIndex == that.readIndex && writeIndex == that.writeIndex && Objects.deepEquals(nanoTimeRingBuffer,
+                    that.nanoTimeRingBuffer) && Objects.deepEquals(scoreRingBuffer, that.scoreRingBuffer);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(readIndex, writeIndex, Arrays.hashCode(nanoTimeRingBuffer), Arrays.hashCode(scoreRingBuffer));
+        }
+
+        @Override
+        public String toString() {
+            return "RingBufferState{" +
+                    "readIndex=" + readIndex +
+                    ", writeIndex=" + writeIndex +
+                    ", nanoTimeRingBuffer=" + Arrays.toString(nanoTimeRingBuffer) +
+                    ", scoreRingBuffer=" + Arrays.toString(scoreRingBuffer) +
+                    '}';
+        }
+    }
+
+    @NonNull
+    RingBufferState getState() {
+        return new RingBufferState(readIndex, writeIndex, nanoTimeRingBuffer, scoreRingBuffer);
+    }
+
+    void resize() {
+        var newCapacity = nanoTimeRingBuffer.length * 2;
+        var newNanoTimeRingBuffer = new long[newCapacity];
+        @SuppressWarnings("unchecked")
+        var newScoreRingBuffer = (Score_[]) new Score[newCapacity];
+
+        if (readIndex < writeIndex) {
+            // entries are [startIndex, writeIndex)
+            var newLength = writeIndex - readIndex;
+            System.arraycopy(nanoTimeRingBuffer, readIndex, newNanoTimeRingBuffer, 0, newLength);
+            System.arraycopy(scoreRingBuffer, readIndex, newScoreRingBuffer, 0, newLength);
+            readIndex = 0;
+            writeIndex = newLength;
+        } else {
+            // entries are [readIndex, CAPACITY) + [0, writeIndex)
+            var firstLength = nanoTimeRingBuffer.length - readIndex;
+            var secondLength = writeIndex;
+            var totalLength = firstLength + secondLength;
+
+            System.arraycopy(nanoTimeRingBuffer, readIndex, newNanoTimeRingBuffer, 0, firstLength);
+            System.arraycopy(scoreRingBuffer, readIndex, newScoreRingBuffer, 0, firstLength);
+            System.arraycopy(nanoTimeRingBuffer, 0, newNanoTimeRingBuffer, firstLength, secondLength);
+            System.arraycopy(scoreRingBuffer, 0, newScoreRingBuffer, firstLength, secondLength);
+            readIndex = 0;
+            writeIndex = totalLength;
+        }
+        nanoTimeRingBuffer = newNanoTimeRingBuffer;
+        scoreRingBuffer = newScoreRingBuffer;
+    }
+
+    public void clear() {
```

Comment:
I wonder if:

- Throwing away the array and re-allocating it would be faster than clearing it. (Maybe not.)
- We should allocate the array using the original size, not using whatever size it is now.

Reply:
We should keep the current size; the size of this buffer is based on score calculation speed (the higher the speed, the more entries in the buffer). Since score calculation speed stays roughly the same in LS, the buffer has an ideal size that it will always regrow to if was shrunk.

Creating a new array each time creates additional memory pressure (although clearing should be rare in most cases) and may invalidate the CPU cache.